File: gamebet_backend/esport-api/default_odds.js
"use strict";

const DEFAULT_ODDS_FILE = "default_odds";

function createDefaultOddsApi(readJson, writeJson) {
  function readStore() {
    const raw = readJson(DEFAULT_ODDS_FILE, {});
    return raw && typeof raw === "object" ? raw : {};
  }

  function writeStore(store) {
    writeJson(DEFAULT_ODDS_FILE, store);
  }

  /** 从合并列表各盘 Sources 取主客最大赔，用于快照回退 */
  function snapshotFromBets(bets) {
    const out = {};
    for (const bet of bets || []) {
      let home = 0;
      let away = 0;
      for (const src of Object.values(bet.Sources || {})) {
        home = Math.max(home, Number(src.HomeOdds) || 0);
        away = Math.max(away, Number(src.AwayOdds) || 0);
      }
      const id = Number(bet.ID);
      if (!id) continue;
      if (home > 0) out[`${id}:Home`] = home;
      if (away > 0) out[`${id}:Away`] = away;
    }
    return out;
  }

  /** 首次见到的赔率写入 default_odds.json（只增不改） */
  function recordFromMatchList(matches) {
    const store = readStore();
    let changed = false;
    for (const match of matches || []) {
      const snap = snapshotFromBets(match.Bets);
      for (const [key, odds] of Object.entries(snap)) {
        if (odds > 0 && !store[key]) {
          store[key] = odds;
          changed = true;
        }
      }
    }
    if (changed) writeStore(store);
    return store;
  }

  function getMatchDefaultOdds(matchIds, buildMatchList) {
    const wanted = new Set((matchIds || []).map((id) => Number(id)).filter(Boolean));
    const out = {};
    if (!wanted.size) return out;

    const store = readStore();
    const matches = buildMatchList();
    for (const match of matches) {
      if (!wanted.has(Number(match.ID))) continue;
      const snap = snapshotFromBets(match.Bets);
      for (const [key, odds] of Object.entries(snap)) {
        out[key] = store[key] > 0 ? store[key] : odds;
      }
    }
    return out;
  }

  function getDefaultOddsSingle(betId, team, buildMatchList) {
    const key = `${Number(betId)}:${team}`;
    const store = readStore();
    if (store[key] > 0) return store[key];

    const matches = buildMatchList();
    for (const match of matches) {
      for (const bet of match.Bets || []) {
        if (Number(bet.ID) !== Number(betId)) continue;
        const snap = snapshotFromBets([bet]);
        return snap[key] || 0;
      }
    }
    return 0;
  }

  return {
    recordFromMatchList,
    getMatchDefaultOdds,
    getDefaultOddsSingle,
    readStore,
  };
}

module.exports = { createDefaultOddsApi, DEFAULT_ODDS_FILE };

File: gamebet_backend_python/esport_api/default_odds.py
from esport_api import store as _store

_FILE = "default_odds"


def _read() -> dict:
    raw = _store.read_json(_FILE, {})
    return raw if isinstance(raw, dict) else {}


def _write(data: dict):
    _store.write_json(_FILE, data)


def record_from_match_list(matches: list):
    store = _read()
    changed = False
    for match in matches or []:
        for bet in match.get("Bets") or []:
            bid = bet.get("ID")
            if not bid:
                continue
            for src in (bet.get("Sources") or {}).values():
                home = float(src.get("HomeOdds") or 0)
                away = float(src.get("AwayOdds") or 0)
                if home > 0 and f"{bid}:Home" not in store:
                    store[f"{bid}:Home"] = home
                    changed = True
                if away > 0 and f"{bid}:Away" not in store:
                    store[f"{bid}:Away"] = away
                    changed = True
    if changed:
        _write(store)


def get_match_default_odds(match_ids: list, build_match_list_fn) -> dict:
    store = _read()
    result = {}
    match_list = None
    for mid in match_ids or []:
        key_h = f"{mid}:Home"
        key_a = f"{mid}:Away"
        if key_h in store or key_a in store:
            result[str(mid)] = {
                "Home": store.get(key_h, 0),
                "Away": store.get(key_a, 0),
            }
        else:
            if match_list is None:
                match_list = build_match_list_fn()
            for m in match_list:
                for bet in m.get("Bets") or []:
                    if str(bet.get("ID")) == str(mid):
                        for src in (bet.get("Sources") or {}).values():
                            result[str(mid)] = {
                                "Home": float(src.get("HomeOdds") or 0),
                                "Away": float(src.get("AwayOdds") or 0),
                            }
    return result


def get_default_odds_single(bet_id: int, team: str, build_match_list_fn) -> float:
    store = _read()
    key = f"{bet_id}:{team}"
    if key in store:
        return float(store[key])
    match_list = build_match_list_fn()
    for m in match_list:
        for bet in m.get("Bets") or []:
            if int(bet.get("ID") or 0) == bet_id:
                for src in (bet.get("Sources") or {}).values():
                    odds_key = "HomeOdds" if team == "Home" else "AwayOdds"
                    return float(src.get(odds_key) or 0)
    return 0.0
